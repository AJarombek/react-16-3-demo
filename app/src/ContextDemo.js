/**
 * Describe the context API introduced in React 16.3.
 * @author Andrew Jarombek
 * @since 12/30/2019
 */

import React, { useContext } from 'react';
import { useHistory } from 'react-router-dom';
import { AJNavTextCircle, AJCodeSnippet } from 'jarombek-react-components';
import FeaturePage from './FeaturePage';
import ThemeWithProps from './context/ThemeWithProps';
import ThemeWithContext from './context/ThemeWithContext';

const themeWithPropsCode =
`// ThemeWithProps.js

import React, {useState} from 'react';
import {useHistory} from 'react-router-dom';
import {AJTextCard, AJSwitchIcon} from 'jarombek-react-components';
import classnames from 'classnames';

import green from './assets/green.png';
import light from './assets/light.png';

const ThemeWithProps = () => {
  const [theme, setTheme] = useState('light');

  return (
    <div className={classnames("theme-with-props", \`\${theme}-theme-with-props\`)}>
      <AJSwitchIcon
        initialState={true}
        disabled={false}
        onChange={() => setTheme(theme === 'dark' ? 'light' : 'dark')}
        offImageUrl={green}
        onImageUrl={light}
      />
      <ThemeWithPropsCard theme={theme} />
    </div>
  );
};
`;

const themeWithPropsCardCode =
`const ThemeWithPropsCard = ({theme}) => {
  return (
    <div className={classnames('theme-with-props-card', \`$\{theme}-theme-with-props-card\`)}>
      <AJTextCard
        title="Theme implemented with props"
        content="... content ..."
        action={null}
        actionText="Button"
        actionDisabled={false}
      />
    </div>
  );
};
`;

const themeWithContextCode =
`// ThemeWithContext.js

import React, { createContext, useContext, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { AJSwitchIcon, AJTextCard } from 'jarombek-react-components';
import classnames from 'classnames';

import green from './assets/green.png';
import light from './assets/light.png';

const ThemeContext = createContext('light');

const ThemeWithContext = () => {
  const [theme, setTheme] = useState('light');

  return (
    <div className={classnames("theme-with-context", \`\${theme}-theme-with-context\`)}>
      <ThemeContext.Provider value={theme}>
        <AJSwitchIcon
          initialState={true}
          disabled={false}
          onChange={() => setTheme(theme === 'dark' ? 'light' : 'dark')}
          offImageUrl={green}
          onImageUrl={light}
        />
        <ThemeWithContextCard />
      </ThemeContext.Provider>
    </div>
  );
};
`;

const themeWithContextCardCode =
`const ThemeWithContextCard = () => {
  const context = useContext(ThemeContext);

  return (
    <div className={classnames('theme-with-context-card', \`\${context}-theme-with-context-card\`)}>
      <AJTextCard
        title="Theme implemented with the context API"
        content="... content ..."
        action={null}
        actionText="Button"
        actionDisabled={false}
      />
    </div>
  );
};
`;

const ContextDemo = () => {
  return (
    <FeaturePage>
      <h1>Context API</h1>
      <div className="demo-body">
        <p>
          In React 16.3, the Context API was rewritten. Context allows data to be shared
          amongst components without passing it via props from parents to children<sup>1</sup>.
          While I'm still learning which scenarios are appropriate for the Context API, the React
          blog recommends to use it for "global" data used by a tree of React
          components<sup>2</sup>.
        </p>
        <p>
          One good use case for the Context API is setting the stylistic theme of a web page based
          on a toggleable switch.  All the components on a page should be aware of the theme chosen,
          but ideally it doesn't need to be passed to every component via a prop.
        </p>
        <p>
          First I implemented this scenario using props.
        </p>
        <ThemeWithProps />
        <p>
          Next I implemented the same component with the Context API.
        </p>
        <ThemeWithContext />
        <p>
          Both components are functionally equivalent.  To identify the differences between the two
          components, let's take a look at the the code.  First, here is the code for a theme
          configured with props:
        </p>
        <AJCodeSnippet language="javascript">
          {themeWithPropsCode}
        </AJCodeSnippet>
        <AJCodeSnippet>
          {themeWithPropsCardCode}
        </AJCodeSnippet>
        <p>
          Next, here is the code for a theme configured with the Context API:
        </p>
        <AJCodeSnippet language="javascript">
          {themeWithContextCode}
        </AJCodeSnippet>
        <AJCodeSnippet>
          {themeWithContextCardCode}
        </AJCodeSnippet>
        <p>
          If you want to see the full code, <a href="https://bit.ly/35Mto2s">
          ThemeWithProps.js</a> and <a href="https://bit.ly/2ThDS7q">ThemeWithContext.js</a> are
          available on GitHub.  In these code samples, its important to focus on how the
          <code>theme</code> variable is passed down to the <code>ThemeWithPropsCard</code> and
          <code>ThemeWithContextCard</code> components.
        </p>
        <p>
          <code>ThemeWithPropsCard</code> receives the selected theme via a prop named
          <code>theme</code>.  The parent component explicitly passes the theme to the child
          component.  In this scenario passing the theme to a single child wasn't much of a hassle.
          However, imagine a chain of nested components, all of which must forward the
          <code>theme</code> prop down to <code>ThemeWithPropsCard</code>.  Components in the middle
          of this chain may not even use the <code>theme</code> prop, but need to be aware of it
          because a nested child needs it.
        </p>
        <p>
          Ideally components only contain code that helps them complete their job.  Creating this
          separation of concerns is possible with the Context API.  With the Context API, global
          variables are created and provided for component chains.  Components can consume the
          context if they need their values, or can just ignore them.  In the previous hypothetical
          example I presented, Components in the middle of the chain don't need to be aware of the
          context they exist in.
        </p>
        <p>
          In <strong>ThemeWithContext.js</strong> I created a context by calling
          <code>createContext()</code>, which is a method imported from React<sup>3</sup>.  The
          argument passed to <code>createContext()</code> is the default value of the context when a
          component does not have a context provider above it in the component hierarchy.  In my
          example, the default value of the context is <code>'light'</code>, so by default
          components should display their light theme.
        </p>
        <p>
          A context provider is a React component attached to a context object as a property.
          Remember that a context object is created with <code>createContext()</code>.  In my
          example, the context object is <code>ThemeContext</code>.  Context provider's are a
          context object's <code>Provider</code> property.  In my example, the context provider of
          <code>ThemeContext</code> is <code>ThemeContext.Provider</code>.
        </p>
        <p>
          All components that are children of a context provider can subscribe to the value it
          provides.  In my example, the <code>ThemeWithContextCard</code> component subscribes to
          the context with the <code>useContext()</code> hook method (Note: Hooks are a React 16.8
          feature.  If you are using an older version of react, you can use
          <code>Context.Consumer</code> for functional components and <code>this.context</code> for
          ES6 class components)<sup>4</sup>.
        </p>
        <p>
          <code>useContext()</code> takes a context object as an argument and returns the current
          value of the context<sup>5</sup>.  <code>ThemeWithContextCard</code> takes the context
          value and uses it to determine the HTML class attached to the component.  This class is
          what toggles the component between light and dark themes.
        </p>
        <p>
          Although my example is very basic, it begins to show the power of the Context API.
          Context really shines in more complex component hierarchies by removing long chains of
          prop exchanges.
        </p>
      </div>
    </FeaturePage>
  );
};

export default ContextDemo;
