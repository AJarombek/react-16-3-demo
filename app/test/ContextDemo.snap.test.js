/**
 * Snapshot test for the {@link ../src/StrictModeComponent} component.
 * @author Andrew Jarombek
 * @since 1/30/2020
 */

import React from 'react';
import ContextDemo from '../src/ContextDemo';
import renderer from 'react-test-renderer';

jest.mock('react-router-dom', () => {
  const historyObj = {
    push: jest.fn()
  };

  return {
    ...jest.requireActual('react-router-dom'),
    useHistory: () => historyObj
  }
});

it('renders correctly', () => {
  const tree = renderer.create(<ContextDemo />).toJSON();
  expect(tree).toMatchSnapshot();
});
