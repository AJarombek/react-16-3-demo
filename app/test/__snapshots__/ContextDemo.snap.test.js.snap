// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="ajNavTextCircle-0-2-1 textRight-0-2-2"
  >
    <div
      className="aj-nav-circle ajNavCircle-0-2-5 ajNavCircle-d0-0-2-13 active-0-2-6"
      onClick={[Function]}
    >
      <figure>
        <img
          className="nav-left left-0-2-11 img-0-2-8 img-d1-0-2-14"
          src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAAXNSR0IArs4c6QAAAVlpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IlhNUCBDb3JlIDUuNC4wIj4KICAgPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4KICAgICAgPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIKICAgICAgICAgICAgeG1sbnM6dGlmZj0iaHR0cDovL25zLmFkb2JlLmNvbS90aWZmLzEuMC8iPgogICAgICAgICA8dGlmZjpPcmllbnRhdGlvbj4xPC90aWZmOk9yaWVudGF0aW9uPgogICAgICA8L3JkZjpEZXNjcmlwdGlvbj4KICAgPC9yZGY6UkRGPgo8L3g6eG1wbWV0YT4KTMInWQAAAjVJREFUeAHt21tqhDAYhmFvu4ki3VnvuugpMgvp4Q/lAxuiJh5iDq8Q0hkdTZ+3zkXBYWBDAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAIHhBYPDAqcZvtpSHjY+Di+p3xO8268+2Xg7SuBifNr4sfFtgyiGkLi5GF82nOHTxu4o8xjuZEQxhMRtHkOGu6KEYuiE3ClxVUIxZJgcZS0Id8p2kLUYzi85iLskUbbhQ0dcEkMXIook4uZLY2gJRJHE+pwlhpZAFEmE56wxtASiSOL/fEsMLYEokvibb42hpRCloBhEKTBG71GK+JpSBH/u7eur6BiK00uUKmL0EqWqGK1HqTJGq1GqjtFalCZitBKlqRi1R2kyRq1Rmo5RW5QuYtQSpasYpUfpMkapUbqOUVoUYqiIzXf/Q5IYsxj68a4oxFCBwJw7CjECEfy3ckUhhi+/8vrqKMRYwV/adVUUYiyJR7x/dhRiRKBvHXJWFGJsSSfsPxqFGAnYsYfujUKMWOEdx6VGIcYO5NSPxEYhRqrsgeNH++xkQw9N+rN77FiPHvv73OvJxmiD7USBrTslFMK997Sx+9nwE9ff5KlSoxAjw59BbBRiZIihS2xFIYakMs5LUYiRMYJ/KT8KMXyhG14rCjFuwF+65Gg73GBDAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAIGSBX4BK4DEtcIi4HkAAAAASUVORK5CYII="
        />
      </figure>
    </div>
    <div
      className="description-0-2-4"
    >
      Home
    </div>
  </div>
  <h1>
    Context API
  </h1>
  <div
    className="demo-body"
  >
    <p>
      In React 16.3, the Context API was rewritten. Context allows data to be shared amongst components without passing it via props from parents to children
      <sup>
        1
      </sup>
      . While I'm still learning which scenarios are appropriate for the Context API, the React blog recommends to use it for "global" data used by a tree of React components
      <sup>
        2
      </sup>
      .
    </p>
    <p>
      One good use case for the Context API is setting the stylistic theme of a web page based on a toggleable switch.  All the components on a page should be aware of the theme chosen, but ideally it doesn't need to be passed to every component via a prop.
    </p>
    <p>
      First I implemented this scenario using props.
    </p>
    <div
      className="theme-with-props light-theme-with-props"
    >
      <div
        className="ajSwitchIconContainer-0-2-15 aj-switch-icon"
      >
        <figure
          className="ajSwitchIcon-0-2-16 ajSwitchIconNotSelected-0-2-17 aj-switch-icon-off"
        >
          <img
            src="test-file-stub"
          />
        </figure>
        <div
          className="ajSwitch-0-2-18 ajSwitchActive-0-2-20 aj-switch"
          onClick={[Function]}
        >
          <div
            className="ajSwitchHead-0-2-22 ajSwitchHeadActive-0-2-23 aj-switch-head"
          >
             
          </div>
          <div
            className="ajSwitchTail-0-2-25 ajSwitchTailActive-0-2-26 aj-switch-tail"
          >
             
          </div>
        </div>
        <figure
          className="ajSwitchIcon-0-2-16 aj-switch-icon-on"
        >
          <img
            src="test-file-stub"
          />
        </figure>
      </div>
      <div
        className="theme-with-props-card light-theme-with-props-card"
      >
        <div
          className="ajTextCard-0-2-28 aj-text-card"
        >
          <div
            className="aj-text-card-header"
          >
            <h2
              className="aj-text-card-title"
            >
              Theme implemented with props
            </h2>
            <p
              className="aj-text-card-subtitle"
            />
          </div>
          <div
            className="aj-text-card-body"
          >
            <p
              className="aj-text-card-content"
            >
              ... content ...
            </p>
          </div>
          <div
            className="aj-text-card-footer"
          >
            <div
              className="ajBaseButton-0-2-32 ajBaseButtonEnabled-0-2-33 aj-text-button ajTextButton-0-2-29 ajTextButtonEnabled-0-2-30"
              onClick={null}
            >
              <button
                disabled={false}
                type="button"
              >
                Button
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <p>
      Next I implemented the same component with the Context API.
    </p>
    <div
      className="theme-with-context light-theme-with-context"
    >
      <div
        className="ajSwitchIconContainer-0-2-15 aj-switch-icon"
      >
        <figure
          className="ajSwitchIcon-0-2-16 ajSwitchIconNotSelected-0-2-17 aj-switch-icon-off"
        >
          <img
            src="test-file-stub"
          />
        </figure>
        <div
          className="ajSwitch-0-2-18 ajSwitchActive-0-2-20 aj-switch"
          onClick={[Function]}
        >
          <div
            className="ajSwitchHead-0-2-22 ajSwitchHeadActive-0-2-23 aj-switch-head"
          >
             
          </div>
          <div
            className="ajSwitchTail-0-2-25 ajSwitchTailActive-0-2-26 aj-switch-tail"
          >
             
          </div>
        </div>
        <figure
          className="ajSwitchIcon-0-2-16 aj-switch-icon-on"
        >
          <img
            src="test-file-stub"
          />
        </figure>
      </div>
      <div
        className="theme-with-context-card light-theme-with-context-card"
      >
        <div
          className="ajTextCard-0-2-28 aj-text-card"
        >
          <div
            className="aj-text-card-header"
          >
            <h2
              className="aj-text-card-title"
            >
              Theme implemented with the context API
            </h2>
            <p
              className="aj-text-card-subtitle"
            />
          </div>
          <div
            className="aj-text-card-body"
          >
            <p
              className="aj-text-card-content"
            >
              ... content ...
            </p>
          </div>
          <div
            className="aj-text-card-footer"
          >
            <div
              className="ajBaseButton-0-2-32 ajBaseButtonEnabled-0-2-33 aj-text-button ajTextButton-0-2-29 ajTextButtonEnabled-0-2-30"
              onClick={null}
            >
              <button
                disabled={false}
                type="button"
              >
                Button
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <p>
      Both components are functionally equivalent.  To identify the differences between the two components, let's take a look at the the code.  First, here is the code for a theme configured with props:
    </p>
    <div
      className="aj-code-snippet"
    >
      <code
        className="javascript hljs-0-2-35"
        title="JavaScript"
      >
        // ThemeWithProps.js

import React, {useState} from 'react';
import {useHistory} from 'react-router-dom';
import {AJTextCard, AJSwitchIcon} from 'jarombek-react-components';
import classnames from 'classnames';

import green from './assets/green.png';
import light from './assets/light.png';

const ThemeWithProps = () =&gt; {
  const [theme, setTheme] = useState('light');

  return (
    &lt;div className={classnames("theme-with-props", \`\${theme}-theme-with-props\`)}&gt;
      &lt;AJSwitchIcon
        initialState={true}
        disabled={false}
        onChange={() =&gt; setTheme(theme === 'dark' ? 'light' : 'dark')}
        offImageUrl={green}
        onImageUrl={light}
      /&gt;
      &lt;ThemeWithPropsCard theme={theme} /&gt;
    &lt;/div&gt;
  );
};

      </code>
    </div>
    <div
      className="aj-code-snippet"
    >
      <code
        className="javascript hljs-0-2-35"
        title=""
      >
        const ThemeWithPropsCard = ({theme}) =&gt; {
  return (
    &lt;div className={classnames('theme-with-props-card', \`\${theme}-theme-with-props-card\`)}&gt;
      &lt;AJTextCard
        title="Theme implemented with props"
        content="... content ..."
        action={null}
        actionText="Button"
        actionDisabled={false}
      /&gt;
    &lt;/div&gt;
  );
};

      </code>
    </div>
    <p>
      Next, here is the code for a theme configured with the Context API:
    </p>
    <div
      className="aj-code-snippet"
    >
      <code
        className="javascript hljs-0-2-35"
        title="JavaScript"
      >
        // ThemeWithContext.js

import React, { createContext, useContext, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { AJSwitchIcon, AJTextCard } from 'jarombek-react-components';
import classnames from 'classnames';

import green from './assets/green.png';
import light from './assets/light.png';

const ThemeContext = createContext('light');

const ThemeWithContext = () =&gt; {
  const [theme, setTheme] = useState('light');

  return (
    &lt;div className={classnames("theme-with-context", \`\${theme}-theme-with-context\`)}&gt;
      &lt;ThemeContext.Provider value={theme}&gt;
        &lt;AJSwitchIcon
          initialState={true}
          disabled={false}
          onChange={() =&gt; setTheme(theme === 'dark' ? 'light' : 'dark')}
          offImageUrl={green}
          onImageUrl={light}
        /&gt;
        &lt;ThemeWithContextCard /&gt;
      &lt;/ThemeContext.Provider&gt;
    &lt;/div&gt;
  );
};

      </code>
    </div>
    <div
      className="aj-code-snippet"
    >
      <code
        className="javascript hljs-0-2-35"
        title=""
      >
        const ThemeWithContextCard = () =&gt; {
  const context = useContext(ThemeContext);

  return (
    &lt;div className={classnames('theme-with-context-card', \`\${context}-theme-with-context-card\`)}&gt;
      &lt;AJTextCard
        title="Theme implemented with the context API"
        content="... content ..."
        action={null}
        actionText="Button"
        actionDisabled={false}
      /&gt;
    &lt;/div&gt;
  );
};

      </code>
    </div>
    <p>
      If you want to see the full code, 
      <a
        href="https://bit.ly/35Mto2s"
      >
        ThemeWithProps.js
      </a>
       and 
      <a
        href="https://bit.ly/2ThDS7q"
      >
        ThemeWithContext.js
      </a>
       are available on GitHub.  In these code samples, its important to focus on how the
      <code>
        theme
      </code>
       variable is passed down to the 
      <code>
        ThemeWithPropsCard
      </code>
       and
      <code>
        ThemeWithContextCard
      </code>
       components.
    </p>
    <p>
      <code>
        ThemeWithPropsCard
      </code>
       receives the selected theme via a prop named
      <code>
        theme
      </code>
      .  The parent component explicitly passes the theme to the child component.  In this scenario passing the theme to a single child wasn't much of a hassle. However, imagine a chain of nested components, all of which must forward the
      <code>
        theme
      </code>
       prop down to 
      <code>
        ThemeWithPropsCard
      </code>
      .  Components in the middle of this chain may not even use the 
      <code>
        theme
      </code>
       prop, but need to be aware of it because a nested child needs it.
    </p>
    <p>
      Ideally components only contain code that helps them complete their job.  Creating this separation of concerns is possible with the Context API.  With the Context API, global variables are created and provided for component chains.  Components can consume the context if they need their values, or can just ignore them.  In the previous hypothetical example I presented, Components in the middle of the chain don't need to be aware of the context they exist in.
    </p>
    <p>
      In 
      <strong>
        ThemeWithContext.js
      </strong>
       I created a context by calling
      <code>
        createContext()
      </code>
      , which is a method imported from React
      <sup>
        3
      </sup>
      .  The argument passed to 
      <code>
        createContext()
      </code>
       is the default value of the context when a component does not have a context provider above it in the component hierarchy.  In my example, the default value of the context is 
      <code>
        'light'
      </code>
      , so by default components should display their light theme.
    </p>
    <p>
      A context provider is a React component attached to a context object as a property. Remember that a context object is created with 
      <code>
        createContext()
      </code>
      .  In my example, the context object is 
      <code>
        ThemeContext
      </code>
      .  Context provider's are a context object's 
      <code>
        Provider
      </code>
       property.  In my example, the context provider of
      <code>
        ThemeContext
      </code>
       is 
      <code>
        ThemeContext.Provider
      </code>
      .
    </p>
    <p>
      All components that are children of a context provider can subscribe to the value it provides.  In my example, the 
      <code>
        ThemeWithContextCard
      </code>
       component subscribes to the context with the 
      <code>
        useContext()
      </code>
       hook method (Note: Hooks are a React 16.8 feature.  If you are using an older version of react, you can use
      <code>
        Context.Consumer
      </code>
       for functional components and 
      <code>
        this.context
      </code>
       for ES6 class components)
      <sup>
        4
      </sup>
      .
    </p>
    <p>
      <code>
        useContext()
      </code>
       takes a context object as an argument and returns the current value of the context
      <sup>
        5
      </sup>
      .  
      <code>
        ThemeWithContextCard
      </code>
       takes the context value and uses it to determine the HTML class attached to the component.  This class is what toggles the component between light and dark themes.
    </p>
    <p>
      Although my example is very basic, it begins to show the power of the Context API. Context really shines in more complex component hierarchies by removing long chains of prop exchanges.
    </p>
  </div>
</div>
`;
